### ISSUE TITLE
False Negative: `StringLiteralEqualityCheck` misses `new String(...) == "literal"`

### ISSUE THREAD

**Issue by Carlson-JLQ** [checkstyle/checkstyle#19475]
Version

Checkstyle：13.3.0

I have read check documentation: https://checkstyle.org/checks/coding/stringliteralequality.html#StringLiteralEquality
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac output/codeql/StringComparison/src/main/java/scensct/core/pos/PosCase1.java
#[[MAKE SURE THERE IS SUCCESSFUL COMPILATION]]

/var/tmp $ cat config.xml

<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="StringLiteralEqualityCheck"/>
    </module>
</module>


/var/tmp $ cat output/codeql/StringComparison/src/main/java/scensct/core/pos/PosCase1.java
// Comparing a new String object with a string literal using == should be flagged as a reference equality comparison with a non-interned operand.
package scensct.core.pos;

public class PosCase1 {
    public static void main(String[] args) {
        String literal = "hello";
        String newString = new String("hello");
        // Scenario: new String("literal") is not guaranteed to be interned, compared with ==
        boolean result = (newString == literal); // Should trigger report // [REPORTED LINE]
        System.out.println(result);
    }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-13.3.0-all.jar -c config.xml output/codeql/StringComparison/src/main/java/scensct/core/pos/PosCase1.java

Starting audit...
Audit done.
```

---

**Describe what you expect in detail.**

The official documentation says this check reports string literals used with `==` or `!=`, because those operators compare references rather than string values. This test case does exactly that:

```java
String literal = "hello";
String newString = new String("hello");
boolean result = (newString == literal);
```

The right-hand operand is a string literal, so the code falls squarely within the rule's documented scope. The fact that the left-hand side is `new String("hello")` makes the comparison even more suspicious, because that object is not guaranteed to be the same reference as the interned literal. According to the rule description, this should be reported and the code should use `equals()` instead.

Checkstyle 13.3.0 produces no violation for this file, so this looks like a false negative.

**Comment by VKS1572** [comment]
I would like to work on this issue.

Root cause seems to be that StringLiteralEqualityCheck only checks direct STRING_LITERAL tokens and does not inspect nested AST (e.g., new String("...")).

I plan to enhance detection by recursively checking for STRING_LITERAL nodes inside expressions.

Please let me know if this approach is acceptable.

**Comment by romani** [comment]
this is limitation Of Checkstyle, 
https://checkstyle.org/writingchecks.html#Limitations

> You cannot determine the type of an expression.

and we can not reliably know type of any variable. on some cases we can , but it creates huge complexity for Check implementation. 

in decision to be smart and possibly do false-positive, we do less and ok with false-negatives, as it annoys users less.
